### PR TITLE
Upload

### PR DESCRIPTION
**Pull request by josephgarner**
- Multiple files now upload at the same time
- User can remove files one at a time
- user is notified if the upload is in progress, uploaded, or failed
- currently the upload is triggered on the upload button press however this may be moved to the submit button

### CODE REVIEW

File: src/components/subcomponents/upload.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
+import React, { useState } from "react";
```
Comment (jpeyper) [inline src/components/subcomponents/upload.js:1]: Honestly, this file is huge and has a lot going on. I think we should consider breaking this up with redux.

Pseudo code for a fileUpload thunk
1. fire UPLOAD_STARTED action
2. get upload url
3. start upload
4. track upload progress and fire UPLOAD_PROGRESS actions (https://gist.github.com/virolea/e1af9359fe071f24de3da3500ff0f429)
5. fire UPLOAD_FINISHED action
6. if any errors, fire UPLOAD_ERROR action (we definitely need error handling for file uploads, things go wrong with big files).

File: src/components/subcomponents/upload.js [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                      case "FAILED":
+                        return (
+                          <Typography variant="body2" color="secondary">
+                            Upload Failed
+                          </Typography>
+                        );
+                      default:
+                        return null;
+                    }
+                  })()}
```
Comment (jpeyper) [inline src/components/subcomponents/upload.js:191]: this self invoking function is a bit hidden, can we create a `<FileUploadProgress>` component that handles this outside of this one huge component

**Comment by jpeyper** [comment]
I'm still reviewing this, but i should have mentioned this earlier... https://react-dropzone.js.org/

File: src/components/subcomponents/upload/file-ui.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                    onDelete(index);
+                  }}
+                >
+                  <HighlightOffIcon />
+                </IconButton>
+              );
+            default:
+              return null;
+          }
+        }}
```
[the remark this answers is not in the document]
Comment (josephgarner) [reply src/components/subcomponents/upload/file-ui.js]: Did not know you could do this, I was running into issues using it so I will try this method

File: src/components/subcomponents/upload.js (hunk 2 above)
Comment (josephgarner) [reply src/components/subcomponents/upload.js]: Yep I can move it out

**Comment by HuangStanley050** [comment]
This PR is truly epic

File: src/store/actions/submissionActions.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  return async dispatch => {
+    dispatch(uploadWarmup(fileID));
+  };
+};
+
+export const cancelFileUpload = fileID => {
+  return async dispatch => {
+    dispatch(uploadCanceled(fileID));
+  };
+};
```
Comment (jpeyper) [inline src/store/actions/submissionActions.js:50]: these dont need to be thunks, they are only dispatching a single action and these actions have no need to be async

File: src/store/actions/submissionActions.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      headers: {
+        Authorization: "Bearer " + getState().auth.jwt
+      }
+    };
+    let files = [];
+    for (let file in submitData.files) {
+      files.push({
+        name: submitData.files[file].name,
+        size: submitData.files[file].size,
+        type: submitData.files[file].type
```
Comment (jpeyper) [inline src/store/actions/submissionActions.js:112]: why do we need to tell the backend about these two? why do we need to store them in the db at all? it should tell us after the upload has finished by sourcing the data from s3.

File: src/components/subcomponents/upload/upload.js [hunk 6]
```diff
@@ -1,110 +1,50 @@
...
+    setFiles({ ...files, ...toUpload });
+    e.target.files = null;
   };
-  const onDelete = (file) => {
-    let delete_Array = files
-    delete_Array.splice(files.indexOf(file),1)
-    setFiles([...delete_Array])
+  const onDelete = fileID => {
+    let delete_obj = files;
+    delete delete_obj[fileID];
```
Comment (jpeyper) [inline src/components/subcomponents/upload/upload.js:44]: uh oh, you're mutating your props, if you want mutate the array, make a copy of it 

```
const delete_obj = [...files]
delete delete_obj[fileID];
```

However, I'm wondering why you need this prop at all, doesn't redux hold all this information for you in `uploadFiles`?

File: src/store/actions/submissionActions.js (hunk 4 above)
Comment (josephgarner) [reply src/store/actions/submissionActions.js:50]: Ah yep, these have been removed

File: src/store/actions/submissionActions.js (hunk 5 above)
Comment (josephgarner) [reply src/store/actions/submissionActions.js:116]: The file size is reported by s3 however the type is not. The reason behind telling the backend about this is when the user wants to view the submission details the file size and type can be displayed to them.

Comment (jpeyper) [reply src/store/actions/submissionActions.js:116]: im not saying we should tell the user those details, I'm saying we shouldn't trust the client to give us the right ones. different browsers may give different types for the same type of file (eg, pdfs have been identified as `application/pdf` and `application/x-pdf` over the years, older browsers may still use the x-pdf one).

https://www.npmjs.com/package/file-type or https://www.npmjs.com/package/mime-types might solve the problem of identify the type of file based on the file itself and not the clients lies.

File: src/components/subcomponents/upload/upload.js (hunk 6 above)
Comment (josephgarner) [reply src/components/subcomponents/upload/upload.js]: Yep, had a mix of things. I've cleaned it up now

File: src/store/actions/submissionActions.js (hunk 5 above)
Comment (josephgarner) [reply src/store/actions/submissionActions.js]: Ohh right okay, I was unaware that browsers could report the wrong file type I know different OS's can report different file sizes. I'll make a note of the package however for now I will remove the file type for now as we are not using it in the immediate future.

**Pull request merged**